Client/mods/deathmatch/logic/luadefs/CLua2DFXDefs.cpp

```diff
@@ -0,0 +1,341 @@
+
+#include "StdInc.h"
+#include "CLua2DFXDefs.h"
+#include <lua/CLuaFunctionParser.h>
+
+void CLua2DFXDefs::LoadFunctions()
+{
+    constexpr static const std::pair<const char*, lua_CFunction> functions[]{
+        // Create / destroy functions
+        {"addModel2DFX", ArgumentParser<AddModel2DFX>},
+        {"removeModel2DFX", ArgumentParser<RemoveModel2DFX>},
+        {"resetModel2DFX", ArgumentParser<ResetModel2DFX>},
+
+        // Set functions
+        {"setModel2DFXProperties", ArgumentParser<SetModel2DFXProperties>},
+        {"set2DFXProperties", ArgumentParser<Set2DFXProperties>},
+        {"setModel2DFXProperty", ArgumentParser<SetModel2DFXProperty>},
+        {"set2DFXProperty", ArgumentParser<Set2DFXProperty>},
+        {"setModel2DFXPosition", ArgumentParser<SetModel2DFXPosition>},
+        {"set2DFXPosition", ArgumentParser<Set2DFXPosition>},
+
+        // Get functions
+        {"getModel2DFXProperties", ArgumentParser<GetModel2DFXProperties>},
+        {"get2DFXProperties", ArgumentParser<Get2DFXProperties>},
+        {"getModel2DFXProperty", ArgumentParser<GetModel2DFXProperty>},
+        {"get2DFXProperty", ArgumentParser<Get2DFXProperty>},
+        {"getModel2DFXPosition", ArgumentParser<GetModel2DFXPosition>},
+        {"get2DFXPosition", ArgumentParser<Get2DFXPosition>},
+        {"getModel2DFXCount", ArgumentParser<GetModel2DFXCount>},
+    };
+
+    // Add functions
+    for (const auto& [name, func] : functions)
+        CLuaCFunctions::AddFunction(name, func);
+}
+
+void CLua2DFXDefs::AddClass(lua_State* luaVM)
+{
+    lua_newclass(luaVM);
+
+    lua_classfunction(luaVM, "add", "addModel2DFX");
+
+    lua_classfunction(luaVM, "setProperties", "set2DFXProperties");
+    lua_classfunction(luaVM, "setProperty", "set2DFXProperty");
+    lua_classfunction(luaVM, "setPosition", "set2DFXPosition");
+
+    lua_classfunction(luaVM, "getProperties", "get2DFXProperties");
+    lua_classfunction(luaVM, "getProperty", "get2DFXProperty");
+    lua_classfunction(luaVM, "getPosition", "get2DFXPosition");
+
+    lua_classvariable(luaVM, "properties", "get2DFXProperties", "set2DFXProperties");
+    lua_classvariable(luaVM, "position", "get2DFXPosition", "set2DFXPosition");
+
+    lua_registerclass(luaVM, "2DFX", "Element");
+}
+
+std::variant<bool, CClient2DFX*> CLua2DFXDefs::AddModel2DFX(lua_State* luaVM, std::uint32_t modelID, CVector position, e2dEffectType effectType, effectDataMap effectData)
+{
+    // Only these effects make sense in MTA
+    if (effectType != e2dEffectType::LIGHT && effectType != e2dEffectType::PARTICLE && effectType != e2dEffectType::ROADSIGN &&
+        effectType != e2dEffectType::ESCALATOR && effectType != e2dEffectType::SUN_GLARE)
+        return false;
+
+    if (!CClient2DFXManager::IsValidModel(modelID))
+        throw std::invalid_argument("Invalid model ID");
+
+    const char* error = CClient2DFXManager::IsValidEffectData(effectType, effectData);
+    if (error)
+        throw LuaFunctionError(error);
```

Comment:
this will throw a warning. Is this intended?

Reply:
Yes